lib/util.js

```diff
@@ -850,11 +850,11 @@ var util = {
   /**
    * @api private
    */
-  calculateRetryDelay: function calculateRetryDelay(retryCount, retryDelayOptions) {
+  calculateRetryDelay: function calculateRetryDelay(retryCount, retryDelayOptions, err) {
     if (!retryDelayOptions) retryDelayOptions = {};
     var customBackoff = retryDelayOptions.customBackoff || null;
     if (typeof customBackoff === 'function') {
-      return customBackoff(retryCount);
+      return customBackoff(retryCount, err);
```

Comment:
Seems like when customBackoff returns negative value, the retry will happen immediately. Can you also update [here](https://github.com/aws/aws-sdk-js/blob/0a02cd851b8e1e2bda7c3e1fa2029a24a5d9d728/lib/event_listeners.js#L476-L481)?

Reply:
Thanks @AllanFly120. I saw I missed a few places near `retryDelays`. Updated those and subsequently the event listener portion. Added a couple tests around those, too.

Added another commit https://github.com/aws/aws-sdk-js/pull/2898/commits/5033e5424b44758315a909bbcc5ee9db4854cdbd that I wasn't too sure about, but saw another PR https://github.com/aws/aws-sdk-js/pull/2721 where that made sense to me to do here.

Reply:
This will break the TS build for all TS users using custom delay. The previous one(optoinal `error`) makes more sense. Even though the `error` is actually always available. But it's a smaller issue than breaking those users' compiliation

Reply:
@AllanFly120 woops, some how missed this comment. i ended up putting the optional back on that error type here https://github.com/aws/aws-sdk-js/pull/2898/commits/96083f99b50a705ecbb8d65760d3da0ef77412a1